hibernate-core/src/main/java/org/hibernate/hql/internal/ast/SqlGenerator.java

```diff
@@ -192,13 +188,16 @@ protected void beginFunctionTemplate(AST node, AST nameNode) {
 		// 		METHOD_NAME
 		FunctionNode functionNode = (FunctionNode) node;
 		SQLFunction sqlFunction = functionNode.getSQLFunction();
+
+		outputStack.addFirst( writer );
+
 		if ( sqlFunction == null ) {
 			// if SQLFunction is null we just write the function out as it appears in the hql statement
+			writer = new StandardFunctionArguments();
 			super.beginFunctionTemplate( node, nameNode );
```

Comment:
Previously no new `writer` was created and both `FORMATDATETIME` and `||` functions share the same writer, thus the strange SQL was created.